Cosimulation error with python 3.2

MyHDL does not seem to like running with Python 3 earlier than 3.4:

            # New pipes are not inheritable by default since py 3.4
            if not PY2:
                for p in rt, wt, rf, wf:
    >               os.set_inheritable(p, True)
    E               AttributeError: 'module' object has no attribute 'set_inheritable'
    
    /usr/local/lib/python3.2/dist-packages/myhdl/_Cosimulation.py:61: AttributeError

--- a/cosimulation/modelsim-win/Makefile
+++ b/cosimulation/modelsim-win/Makefile
@@ -1,0 +1,21 @@
+INCS := C:\\modeltech64_10.4\\include
+LIB_PATH := C:\\modeltech64_10.4\\win64
+
+ifneq ($(filter cl%,$(CC)),)
+	CFLAGS := /LD /I$(INCS)
+	LIBFLAGS := $(LIB_PATH)\\mtipli.lib
+else
+	CFLAGS := -static -g -I$(INCS) -o myhdl_vpi.dll
+	LIBFLAGS := -L$(LIB_PATH) -lmtipli
+endif
+
+all: myhdl_vpi.dll
+
+myhdl_vpi.dll: myhdl_vpi.c
+	$(CC) $(CFLAGS) $< $(LIBFLAGS)
+
+clean:
+	@del /q myhdl_vpi.dll || rm -rf myhdl_vpi.dll 2>&1> /dev/null ||:
+	@del /q myhdl_vpi.lib || rm -rf myhdl_vpi.lib 2>&1> /dev/null ||:
+	@del /q myhdl_vpi.exp || rm -rf myhdl_vpi.exp 2>&1> /dev/null ||:
+	@del /q myhdl_vpi.obj || rm -rf myhdl_vpi.obj 2>&1> /dev/null ||:

--- a/cosimulation/modelsim-win/myhdl_vpi.c
+++ b/cosimulation/modelsim-win/myhdl_vpi.c
@@ -1,9 +1,9 @@
 
 #include <stdlib.h>
-#include <unistd.h>
 #include <assert.h>
 #include <string.h>
 #include <stdio.h>
+#include <Windows.h>
 #include "vpi_user.h"
 #include "sv_vpi_user.h"
 
@@ -26,8 +26,8 @@ typedef unsigned char   PLI_UBYTE8;
 /* 64 bit type for time calculations */
 typedef unsigned long long myhdl_time64_t;
 
-static int rpipe;
-static int wpipe;
+static HANDLE rpipe;
+static HANDLE wpipe;
 
 static vpiHandle from_myhdl_systf_handle = NULL;
 static vpiHandle to_myhdl_systf_handle = NULL;
@@ -62,6 +62,30 @@ static myhdl_time64_t timestruct_to_time(const struct t_vpi_time*ts)
       return ti;
 }
 
+static int read_pipe(void *buf, size_t count)
+{
+    DWORD read;
+
+    if (!ReadFile(rpipe, buf, (DWORD)count, &read, NULL))
+    {
+        return -1;
+    }
+
+    return (int)read;
+}
+
+static int write_pipe(const void *buf, size_t count)
+{
+    DWORD written;
+
+    if (!WriteFile(wpipe, buf, (DWORD)count, &written, NULL))
+    {
+        return -1;
+    }
+
+    return (int)written;
+}
+
 static int init_pipes()
 {
   char *w;
@@ -83,8 +107,8 @@ static int init_pipes()
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
   }
-  wpipe = atoi(w);
-  rpipe = atoi(r);
+  wpipe = (HANDLE)atoi(w);
+  rpipe = (HANDLE)atoi(r);
   init_pipes_flag = 1;
   return (0);
 }
@@ -137,9 +161,9 @@ static PLI_INT32 from_myhdl_calltf(PLI_BYTE8 *user_data)
   }
   //vpi_free_object(reg_iter);
 
-  n = write(wpipe, buf, strlen(buf));
+  n = write_pipe(buf, strlen(buf));
 
-  if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) == 0) {
     vpi_printf("Info: MyHDL simulator down\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -216,9 +240,9 @@ static PLI_INT32 to_myhdl_calltf(PLI_BYTE8 *user_data)
   }
   //vpi_free_object(net_iter);
 
-  n = write(wpipe, buf, strlen(buf));
+  n = write_pipe(buf, strlen(buf));
 
-  if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) == 0) {
     vpi_printf("ABORT from $to_myhdl\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -274,9 +298,9 @@ static PLI_INT32 readonly_callback(p_cb_data cb_data)
 
   if (start_flag) {
     start_flag = 0;
-    n = write(wpipe, "START", 5);
+    n = write_pipe("START", 5);
     // vpi_printf("INFO: RO cb at start-up\n");
-    if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+    if ((n = read_pipe(buf, MAXLINE)) == 0) {
       vpi_printf("ABORT from RO cb at start-up\n");
       vpi_control(vpiFinish, 1);  /* abort simulation */
     }
@@ -312,8 +336,8 @@ static PLI_INT32 readonly_callback(p_cb_data cb_data)
   }
   //vpi_free_object(net_iter);
 
-  n = write(wpipe, buf, strlen(buf));
-  if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+  n = write_pipe(buf, strlen(buf));
+  if ((n = read_pipe(buf, MAXLINE)) == 0) {
     // vpi_printf("ABORT from RO cb\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -463,6 +487,7 @@ static PLI_INT32 change_callback(p_cb_data cb_data)
 }
 
 
+__declspec(dllexport)
 void myhdl_register()
 {
   s_vpi_systf_data tf_data;
@@ -486,6 +511,7 @@ void myhdl_register()
   //vpi_free_object(tf_data);               // @mod cfelton
 }
 
+__declspec(dllexport)
 void (*vlog_startup_routines[])() = {
       myhdl_register,
       0
@@ -493,6 +519,7 @@ void (*vlog_startup_routines[])() = {
 
 /* dummy +loadvpi= boostrap routine - mimics old style exec all routines */
 /* in standard PLI vlog_startup_routines table */
+__declspec(dllexport)
 void vpi_compat_bootstrap(void)
 {
  int i;

--- a/cosimulation/modelsim-win/myhdl_vpi.c
+++ b/cosimulation/modelsim-win/myhdl_vpi.c
@@ -163,7 +163,7 @@ static PLI_INT32 from_myhdl_calltf(PLI_BYTE8 *user_data)
 
   n = write_pipe(buf, strlen(buf));
 
-  if ((n = read_pipe(buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) <= 0) {
     vpi_printf("Info: MyHDL simulator down\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -242,7 +242,7 @@ static PLI_INT32 to_myhdl_calltf(PLI_BYTE8 *user_data)
 
   n = write_pipe(buf, strlen(buf));
 
-  if ((n = read_pipe(buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) <= 0) {
     vpi_printf("ABORT from $to_myhdl\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -300,7 +300,7 @@ static PLI_INT32 readonly_callback(p_cb_data cb_data)
     start_flag = 0;
     n = write_pipe("START", 5);
     // vpi_printf("INFO: RO cb at start-up\n");
-    if ((n = read_pipe(buf, MAXLINE)) == 0) {
+    if ((n = read_pipe(buf, MAXLINE)) <= 0) {
       vpi_printf("ABORT from RO cb at start-up\n");
       vpi_control(vpiFinish, 1);  /* abort simulation */
     }
@@ -337,7 +337,7 @@ static PLI_INT32 readonly_callback(p_cb_data cb_data)
   //vpi_free_object(net_iter);
 
   n = write_pipe(buf, strlen(buf));
-  if ((n = read_pipe(buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) <= 0) {
     // vpi_printf("ABORT from RO cb\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);

--- a/cosimulation/modelsim-win/test/bin2gray.py
+++ b/cosimulation/modelsim-win/test/bin2gray.py
@@ -1,0 +1,13 @@
+import os
+
+from myhdl import Cosimulation
+
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_bin2gray'
+       
+def bin2gray(B, G, width):
+    os.system('vlog -quiet +define+width=%s ../../test/verilog/bin2gray.v' % (width))
+    os.system('vlog -quiet +define+width=%s ../../test/verilog/dut_bin2gray.v' % (width)) 
+
+    return Cosimulation(cmd, B=B, G=G)
+               
+

--- a/cosimulation/modelsim-win/test/bin2gray.py
+++ b/cosimulation/modelsim-win/test/bin2gray.py
@@ -2,12 +2,10 @@ import os
 
 from myhdl import Cosimulation
 
-cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_bin2gray'
-       
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.dll -do cosim.do dut_bin2gray'
+
 def bin2gray(B, G, width):
     os.system('vlog -quiet +define+width=%s ../../test/verilog/bin2gray.v' % (width))
-    os.system('vlog -quiet +define+width=%s ../../test/verilog/dut_bin2gray.v' % (width)) 
+    os.system('vlog -quiet +define+width=%s ../../test/verilog/dut_bin2gray.v' % (width))
 
     return Cosimulation(cmd, B=B, G=G)
-               
-

--- a/cosimulation/modelsim-win/test/cosim.do
+++ b/cosimulation/modelsim-win/test/cosim.do
@@ -1,0 +1,2 @@
+run -all
+quit

--- a/cosimulation/modelsim-win/test/dff.py
+++ b/cosimulation/modelsim-win/test/dff.py
@@ -1,0 +1,12 @@
+import os
+
+from myhdl import Cosimulation
+
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_dff'
+      
+def dff(q, d, clk, reset):
+    os.system('vlog -quiet ../../test/verilog/dff.v')
+    os.system('vlog -quiet ../../test/verilog/dut_dff.v') 
+
+    return Cosimulation(cmd, **locals())
+               

--- a/cosimulation/modelsim-win/test/dff.py
+++ b/cosimulation/modelsim-win/test/dff.py
@@ -2,11 +2,10 @@ import os
 
 from myhdl import Cosimulation
 
-cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_dff'
-      
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.dll -do cosim.do dut_dff'
+
 def dff(q, d, clk, reset):
     os.system('vlog -quiet ../../test/verilog/dff.v')
-    os.system('vlog -quiet ../../test/verilog/dut_dff.v') 
+    os.system('vlog -quiet ../../test/verilog/dut_dff.v')
 
     return Cosimulation(cmd, **locals())
-               

--- a/cosimulation/modelsim-win/test/dff_clkout.py
+++ b/cosimulation/modelsim-win/test/dff_clkout.py
@@ -1,0 +1,13 @@
+import os
+import os.path as path
+
+from myhdl import Cosimulation
+
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_dff_clkout'
+      
+def dff_clkout(clkout, q, d, clk, reset):
+    os.system('vlog -quiet ../../test/verilog/dff_clkout.v')
+    os.system('vlog -quiet ../../test/verilog/dut_dff_clkout.v')
+    
+    return Cosimulation(cmd, **locals())
+               

--- a/cosimulation/modelsim-win/test/dff_clkout.py
+++ b/cosimulation/modelsim-win/test/dff_clkout.py
@@ -1,13 +1,11 @@
 import os
-import os.path as path
 
 from myhdl import Cosimulation
 
-cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_dff_clkout'
-      
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.dll -do cosim.do dut_dff_clkout'
+
 def dff_clkout(clkout, q, d, clk, reset):
     os.system('vlog -quiet ../../test/verilog/dff_clkout.v')
     os.system('vlog -quiet ../../test/verilog/dut_dff_clkout.v')
-    
+
     return Cosimulation(cmd, **locals())
-               

--- a/cosimulation/modelsim-win/test/inc.py
+++ b/cosimulation/modelsim-win/test/inc.py
@@ -1,0 +1,12 @@
+import os
+
+from myhdl import Cosimulation
+
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_inc'
+
+def inc(count, enable, clock, reset, n):
+    os.system('vlog -quiet +define+n=%s ../../test/verilog/inc.v' % (n))
+    os.system('vlog -quiet +define+n=%s ../../test/verilog/dut_inc.v' % (n)) 
+
+    return Cosimulation(cmd, **locals())
+               

--- a/cosimulation/modelsim-win/test/inc.py
+++ b/cosimulation/modelsim-win/test/inc.py
@@ -2,11 +2,10 @@ import os
 
 from myhdl import Cosimulation
 
-cmd = 'vsim -c -quiet -pli ../myhdl_vpi.so -do cosim.do dut_inc'
+cmd = 'vsim -c -quiet -pli ../myhdl_vpi.dll -do cosim.do dut_inc'
 
 def inc(count, enable, clock, reset, n):
     os.system('vlog -quiet +define+n=%s ../../test/verilog/inc.v' % (n))
-    os.system('vlog -quiet +define+n=%s ../../test/verilog/dut_inc.v' % (n)) 
+    os.system('vlog -quiet +define+n=%s ../../test/verilog/dut_inc.v' % (n))
 
     return Cosimulation(cmd, **locals())
-               

--- a/cosimulation/modelsim-win/test/test_all.py
+++ b/cosimulation/modelsim-win/test/test_all.py
@@ -1,0 +1,47 @@
+#  This file is part of the myhdl library, a Python package for using
+#  Python as a Hardware Description Language.
+#
+#  Copyright (C) 2003-2008 Jan Decaluwe
+#
+#  The myhdl library is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU Lesser General Public License as
+#  published by the Free Software Foundation; either version 2.1 of the
+#  License, or (at your option) any later version.
+#
+#  This library is distributed in the hope that it will be useful, but
+#  WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library; if not, write to the Free Software
+#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+
+""" Run cosimulation unit tests. """
+
+
+import sys
+
+sys.path.append("../../test")
+
+import test_bin2gray, test_inc, test_dff
+
+# modules = (test_dff,  )
+modules = (test_bin2gray, test_inc, test_dff )
+
+import unittest
+
+tl = unittest.defaultTestLoader
+def suite():
+    alltests = unittest.TestSuite()
+    for m in modules:
+        alltests.addTest(tl.loadTestsFromModule(m))
+    return alltests
+
+def main():
+    unittest.main(defaultTest='suite',
+                  testRunner=unittest.TextTestRunner(verbosity=2))
+    
+
+if __name__ == '__main__':
+    main()

--- a/cosimulation/modelsim-win/test/test_all.py
+++ b/cosimulation/modelsim-win/test/test_all.py
@@ -41,7 +41,7 @@ def suite():
 def main():
     unittest.main(defaultTest='suite',
                   testRunner=unittest.TextTestRunner(verbosity=2))
-    
+
 
 if __name__ == '__main__':
     main()

--- a/cosimulation/modelsim/myhdl_vpi.c
+++ b/cosimulation/modelsim/myhdl_vpi.c
@@ -1,11 +1,24 @@
 
+#ifndef _WIN32
+#   include <unistd.h>
+#else
+#   include <Windows.h>
+#endif
+
 #include <stdlib.h>
-#include <unistd.h>
 #include <assert.h>
 #include <string.h>
 #include <stdio.h>
 #include "vpi_user.h"
 #include "sv_vpi_user.h"
+
+#if (defined WIN32 || defined _WIN32 || defined WINCE || defined __CYGWIN__)
+#  define LIBRARY_EXPORTS __declspec(dllexport)
+#elif defined __GNUC__ && __GNUC__ >= 4
+#  define LIBRARY_EXPORTS __attribute__ ((visibility ("default")))
+#else
+#  define LIBRARY_EXPORTS
+#endif
 
 #define MAXLINE 4096
 #define MAXWIDTH 10
@@ -26,8 +39,12 @@ typedef unsigned char   PLI_UBYTE8;
 /* 64 bit type for time calculations */
 typedef unsigned long long myhdl_time64_t;
 
-static int rpipe;
-static int wpipe;
+#ifndef _WIN32
+typedef int HANDLE;
+#endif
+
+static HANDLE rpipe;
+static HANDLE wpipe;
 
 static vpiHandle from_myhdl_systf_handle = NULL;
 static vpiHandle to_myhdl_systf_handle = NULL;
@@ -62,6 +79,38 @@ static myhdl_time64_t timestruct_to_time(const struct t_vpi_time*ts)
       return ti;
 }
 
+static int read_pipe(void *buf, size_t count)
+{
+#ifdef _WIN32
+    DWORD read;
+
+    if (!ReadFile(rpipe, buf, (DWORD)count, &read, NULL))
+    {
+        return -1;
+    }
+
+    return (int)read;
+#else
+    return (int)read(rpipe, buf, count);
+#endif
+}
+
+static int write_pipe(const void *buf, size_t count)
+{
+#ifdef _WIN32
+    DWORD written;
+
+    if (!WriteFile(wpipe, buf, (DWORD)count, &written, NULL))
+    {
+        return -1;
+    }
+
+    return (int)written;
+#else
+    return (int)write(wpipe, buf, count);
+#endif
+}
+
 static int init_pipes()
 {
   char *w;
@@ -83,8 +132,8 @@ static int init_pipes()
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
   }
-  wpipe = atoi(w);
-  rpipe = atoi(r);
+  wpipe = (HANDLE)atoi(w);
+  rpipe = (HANDLE)atoi(r);
   init_pipes_flag = 1;
   return (0);
 }
@@ -125,7 +174,7 @@ static PLI_INT32 from_myhdl_calltf(PLI_BYTE8 *user_data)
   while ((reg_handle = vpi_scan(reg_iter)) != NULL) {
     if (vpi_get(vpiType, reg_handle) != vpiReg) {
       vpi_printf("ERROR: $from_myhdl argument %s should be a reg\n",
-		 vpi_get_str(vpiName, reg_handle));
+                vpi_get_str(vpiName, reg_handle));
       vpi_control(vpiFinish, 1);  /* abort simulation */
       return(0);
     }
@@ -137,9 +186,9 @@ static PLI_INT32 from_myhdl_calltf(PLI_BYTE8 *user_data)
   }
   //vpi_free_object(reg_iter);
 
-  n = write(wpipe, buf, strlen(buf));  
+  n = write_pipe(buf, strlen(buf));
 
-  if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) == 0) {
     vpi_printf("Info: MyHDL simulator down\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -216,9 +265,9 @@ static PLI_INT32 to_myhdl_calltf(PLI_BYTE8 *user_data)
   }
   //vpi_free_object(net_iter);
 
-  n = write(wpipe, buf, strlen(buf));
+  n = write_pipe(buf, strlen(buf));
 
-  if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+  if ((n = read_pipe(buf, MAXLINE)) == 0) {
     vpi_printf("ABORT from $to_myhdl\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -274,12 +323,12 @@ static PLI_INT32 readonly_callback(p_cb_data cb_data)
 
   if (start_flag) {
     start_flag = 0;
-    n = write(wpipe, "START", 5);  
+    n = write_pipe("START", 5);
     // vpi_printf("INFO: RO cb at start-up\n");
-    if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+    if ((n = read_pipe(buf, MAXLINE)) == 0) {
       vpi_printf("ABORT from RO cb at start-up\n");
       vpi_control(vpiFinish, 1);  /* abort simulation */
-    }  
+    }
     assert(n > 0);
   }
 
@@ -311,9 +360,9 @@ static PLI_INT32 readonly_callback(p_cb_data cb_data)
     vpi_free_object(net_handle);  // done with this one
   }
   //vpi_free_object(net_iter);
-  
-  n = write(wpipe, buf, strlen(buf));
-  if ((n = read(rpipe, buf, MAXLINE)) == 0) {
+
+  n = write_pipe(buf, strlen(buf));
+  if ((n = read_pipe(buf, MAXLINE)) == 0) {
     // vpi_printf("ABORT from RO cb\n");
     vpi_control(vpiFinish, 1);  /* abort simulation */
     return(0);
@@ -463,7 +512,7 @@ static PLI_INT32 change_callback(p_cb_data cb_data)
 }
 
 
-
+LIBRARY_EXPORTS
 void myhdl_register()
 {
   s_vpi_systf_data tf_data;
@@ -487,6 +536,7 @@ void myhdl_register()
   //vpi_free_object(tf_data);               // @mod cfelton
 }
 
+LIBRARY_EXPORTS
 void (*vlog_startup_routines[])() = {
       myhdl_register,
       0
@@ -494,13 +544,14 @@ void (*vlog_startup_routines[])() = {
 
 /* dummy +loadvpi= boostrap routine - mimics old style exec all routines */
 /* in standard PLI vlog_startup_routines table */
+LIBRARY_EXPORTS
 void vpi_compat_bootstrap(void)
 {
  int i;
 
- for (i = 0;; i++) 
+ for (i = 0;; i++)
   {
-   if (vlog_startup_routines[i] == NULL) break; 
+   if (vlog_startup_routines[i] == NULL) break;
    vlog_startup_routines[i]();
   }
 }

--- a/myhdl/_Cosimulation.py
+++ b/myhdl/_Cosimulation.py
@@ -24,14 +24,14 @@ import sys
 import os
 import shlex
 import subprocess
-if sys.platform == "win32":
-    import msvcrt
 
 from myhdl._intbv import intbv
 from myhdl import _simulator, CosimulationError
 from myhdl._compat import string_types, to_bytes, to_str
+from myhdl._util import _setInheritable
 
 _MAXLINE = 4096
+
 
 class _error:
     pass
@@ -42,6 +42,7 @@ _error.TimeZero = "myhdl vpi call when not at time 0"
 _error.NoCommunication = "No signals communicating to myhdl"
 _error.SimulationEnd = "Premature simulation end"
 _error.OSError = "OSError"
+
 
 class Cosimulation(object):
 
@@ -58,17 +59,13 @@ class Cosimulation(object):
         rt, wt = os.pipe()
         rf, wf = os.pipe()
 
-        # Support for python 3.4 non-inheritable pipes
-        # On < 3.4 Windows still has non-inheritable pipes, but < 3.4 support
-        # was dropped in [#38](https://github.com/jandecaluwe/myhdl/issues/38)
-        if hasattr(os, 'set_inheritable'):
-            # Disable inheritance for ends that we don't want the child to have
-            os.set_inheritable(rt, False)
-            os.set_inheritable(wf, False)
+        # Disable inheritance for ends that we don't want the child to have
+        _setInheritable(rt, False)
+        _setInheritable(wf, False)
 
-            # Enable inheritance for child ends
-            os.set_inheritable(wt, True)
-            os.set_inheritable(rf, True)
+        # Enable inheritance for child ends
+        _setInheritable(wt, True)
+        _setInheritable(rf, True)
 
         self._rt = rt
         self._wf = wf
@@ -91,6 +88,7 @@ class Cosimulation(object):
             env['MYHDL_TO_PIPE'] = str(wt)
             env['MYHDL_FROM_PIPE'] = str(rf)
         else:
+            import msvcrt
             env['MYHDL_TO_PIPE'] = str(msvcrt.get_osfhandle(wt))
             env['MYHDL_FROM_PIPE'] = str(msvcrt.get_osfhandle(rf))
 

--- a/myhdl/_Cosimulation.py
+++ b/myhdl/_Cosimulation.py
@@ -55,44 +55,20 @@ class Cosimulation(object):
             raise CosimulationError(_error.MultipleCosim)
         _simulator._cosim = 1
 
-        if hasattr(os, 'set_inheritable'):
-            set_inheritable = os.set_inheritable
-        else:
-            # No luck on python < 3.4, so we need to implement it ouselves
-            def set_inheritable(fd, inheritable):
-                if sys.platform == "win32":
-                    import ctypes
-
-                    HANDLE_FLAG_INHERIT = 1
-                    flag = 1 if inheritable else 0
-
-                    if ctypes.windll.kernel32.SetHandleInformation(msvcrt.get_osfhandle(fd),
-                                                                   HANDLE_FLAG_INHERIT,
-                                                                   flag) == 0:
-                        raise IOError("Failed on HANDLE_FLAG_INHERIT")
-                else:
-                    import fcntl
-
-                    fd_flags = fcntl.fcntl(fd, fcntl.F_GETFD)
-
-                    if inheritable:
-                        fd_flags &= ~fcntl.FD_CLOEXEC
-                    else:
-                        fd_flags |= fcntl.FD_CLOEXEC
-
-                    fcntl.fcntl(fd, fcntl.F_SETFD, fd_flags)
-
         rt, wt = os.pipe()
         rf, wf = os.pipe()
 
-        # Disable inheritance for ends that we don't want the child to have
-        set_inheritable(rt, False)
-        set_inheritable(wf, False)
+        # Support for python 3.4 non-inheritable pipes
+        # On < 3.4 Windows still has non-inheritable pipes, but < 3.4 support
+        # was dropped in [#38](https://github.com/jandecaluwe/myhdl/issues/38)
+        if hasattr(os, 'set_inheritable'):
+            # Disable inheritance for ends that we don't want the child to have
+            os.set_inheritable(rt, False)
+            os.set_inheritable(wf, False)
 
-        # Enable inheritance for child ends
-        # (needed for python >= 3.4, doesn't hurt for earlier versions)
-        set_inheritable(wt, True)
-        set_inheritable(rf, True)
+            # Enable inheritance for child ends
+            os.set_inheritable(wt, True)
+            os.set_inheritable(rf, True)
 
         self._rt = rt
         self._wf = wf
@@ -109,6 +85,8 @@ class Cosimulation(object):
 
         env = os.environ.copy()
 
+        # In Windows the FDs aren't inheritable when using Popen,
+        # only the HANDLEs are
         if sys.platform != "win32":
             env['MYHDL_TO_PIPE'] = str(wt)
             env['MYHDL_FROM_PIPE'] = str(rf)

--- a/myhdl/_Cosimulation.py
+++ b/myhdl/_Cosimulation.py
@@ -29,7 +29,7 @@ if sys.platform == "win32":
 
 from myhdl._intbv import intbv
 from myhdl import _simulator, CosimulationError
-from myhdl._compat import PY2, string_types, to_bytes, to_str
+from myhdl._compat import string_types, to_bytes, to_str
 
 _MAXLINE = 4096
 
@@ -108,10 +108,13 @@ class Cosimulation(object):
         self._getMode = 1
 
         env = os.environ.copy()
-        env['MYHDL_TO_PIPE'] = str(wt) if sys.platform != "win32"
-                               else str(msvcrt.get_osfhandle(wt))
-        env['MYHDL_FROM_PIPE'] = str(rf) if sys.platform != "win32"
-                                 else str(msvcrt.get_osfhandle(rf))
+
+        if sys.platform != "win32":
+            env['MYHDL_TO_PIPE'] = str(wt)
+            env['MYHDL_FROM_PIPE'] = str(rf)
+        else:
+            env['MYHDL_TO_PIPE'] = str(msvcrt.get_osfhandle(wt))
+            env['MYHDL_FROM_PIPE'] = str(msvcrt.get_osfhandle(rf))
 
         if isinstance(exe, string_types):
             exe = shlex.split(exe)

--- a/myhdl/_Cosimulation.py
+++ b/myhdl/_Cosimulation.py
@@ -20,10 +20,12 @@
 """ Module that provides the Cosimulation class """
 from __future__ import absolute_import
 
-
+import sys
 import os
 import shlex
 import subprocess
+if sys.platform == "win32":
+    import msvcrt
 
 from myhdl._intbv import intbv
 from myhdl import _simulator, CosimulationError
@@ -46,20 +48,54 @@ class Cosimulation(object):
     """ Cosimulation class. """
 
     def __init__(self, exe="", **kwargs):
-        
+
         """ Construct a cosimulation object. """
-        
+
         if _simulator._cosim:
             raise CosimulationError(_error.MultipleCosim)
         _simulator._cosim = 1
-        
-        self._rt, self._wt = rt, wt = os.pipe()
-        self._rf, self._wf = rf, wf = os.pipe()
 
-        # New pipes are not inheritable by default since py 3.4
-        if not PY2:
-            for p in rt, wt, rf, wf:
-                os.set_inheritable(p, True)
+        if hasattr(os, 'set_inheritable'):
+            set_inheritable = os.set_inheritable
+        else:
+            # No luck on python < 3.4, so we need to implement it ouselves
+            def set_inheritable(fd, inheritable):
+                if sys.platform == "win32":
+                    import ctypes
+
+                    HANDLE_FLAG_INHERIT = 1
+                    flag = 1 if inheritable else 0
+
+                    if ctypes.windll.kernel32.SetHandleInformation(msvcrt.get_osfhandle(fd),
+                                                                   HANDLE_FLAG_INHERIT,
+                                                                   flag) == 0:
+                        raise IOError("Failed on HANDLE_FLAG_INHERIT")
+                else:
+                    import fcntl
+
+                    fd_flags = fcntl.fcntl(fd, fcntl.F_GETFD)
+
+                    if inheritable:
+                        fd_flags &= ~fcntl.FD_CLOEXEC
+                    else:
+                        fd_flags |= fcntl.FD_CLOEXEC
+
+                    fcntl.fcntl(fd, fcntl.F_SETFD, fd_flags)
+
+        rt, wt = os.pipe()
+        rf, wf = os.pipe()
+
+        # Disable inheritance for ends that we don't want the child to have
+        set_inheritable(rt, False)
+        set_inheritable(wf, False)
+
+        # Enable inheritance for child ends
+        # (needed for python >= 3.4, doesn't hurt for earlier versions)
+        set_inheritable(wt, True)
+        set_inheritable(rf, True)
+
+        self._rt = rt
+        self._wf = wf
 
         self._fromSignames = fromSignames = []
         self._fromSizes = fromSizes = []
@@ -71,20 +107,17 @@ class Cosimulation(object):
         self._hasChange = 0
         self._getMode = 1
 
-        def close_rt_wf():
-            os.close(rt)
-            os.close(wf)
-
         env = os.environ.copy()
-        env['MYHDL_TO_PIPE'] = str(wt)
-        env['MYHDL_FROM_PIPE'] = str(rf)
+        env['MYHDL_TO_PIPE'] = str(wt) if sys.platform != "win32"
+                               else str(msvcrt.get_osfhandle(wt))
+        env['MYHDL_FROM_PIPE'] = str(rf) if sys.platform != "win32"
+                                 else str(msvcrt.get_osfhandle(rf))
 
         if isinstance(exe, string_types):
             exe = shlex.split(exe)
-            
+
         try:
-            sp = subprocess.Popen(exe, env=env, close_fds=False,
-                                  preexec_fn=close_rt_wf)
+            sp = subprocess.Popen(exe, env=env, close_fds=False)
         except OSError as e:
             raise CosimulationError(_error.OSError, str(e))
 
@@ -154,7 +187,7 @@ class Cosimulation(object):
                 except ValueError:
                     next = intbv(0)
             s.next = next
-                 
+
         self._getMode = 0
 
     def _put(self, time):
@@ -182,7 +215,7 @@ class Cosimulation(object):
         while 1:
             yield sigs
             self._hasChange = 1
-            
+
     def __del__(self):
         """ Clear flag when this object destroyed - to suite unittest. """
         _simulator._cosim = 0

--- a/myhdl/_util.py
+++ b/myhdl/_util.py
@@ -23,14 +23,15 @@
 from __future__ import absolute_import
 from __future__ import print_function
 
-
 import ast
 import sys
+import os
 import inspect
 
 from tokenize import generate_tokens, untokenize, INDENT
 
 from myhdl._compat import integer_types, StringIO
+
 
 def _printExcInfo():
     kind, value = sys.exc_info()[:2]
@@ -91,3 +92,29 @@ def _genfunc(gen):
     else:
         func = gen.genfunc
     return func
+
+if hasattr(os, 'set_inheritable'):
+    _setInheritable = os.set_inheritable
+else:
+    def _setInheritable(fd, inheritable):
+        if sys.platform == "win32":
+            import msvcrt
+            import ctypes.windll.kernel32 as kernel32
+
+            HANDLE_FLAG_INHERIT = 1
+
+            if kernel32.SetHandleInformation(msvcrt.get_osfhandle(fd),
+                                             HANDLE_FLAG_INHERIT,
+                                             1 if inheritable else 0) == 0:
+                raise IOError("Failed on HANDLE_FLAG_INHERIT")
+        else:
+            import fcntl
+
+            fd_flags = fcntl.fcntl(fd, fcntl.F_GETFD)
+
+            if inheritable:
+                fd_flags &= ~fcntl.FD_CLOEXEC
+            else:
+                fd_flags |= fcntl.FD_CLOEXEC
+
+            fcntl.fcntl(fd, fcntl.F_SETFD, fd_flags)

--- a/myhdl/_util.py
+++ b/myhdl/_util.py
@@ -96,6 +96,10 @@ def _genfunc(gen):
 if hasattr(os, 'set_inheritable'):
     _setInheritable = os.set_inheritable
 else:
+# This implementation of set_inheritable is based on a code sample in
+# [PEP 0446](https://www.python.org/dev/peps/pep-0446/) and on the
+# CPython implementation of that proposal which can be browsed [here]
+# (hg.python.org/releasing/3.4/file/8671f89107c8/Modules/posixmodule.c#l11130)
     def _setInheritable(fd, inheritable):
         if sys.platform == "win32":
             import msvcrt

